lib/rule-translation/src/RuleTranslation.hs

```diff
@@ -0,0 +1,449 @@
+    ruleDoc = text "let" <-> (text . showRuleName . L.get preName $ L.get rInfo ru) <-> text "=" $-$
+              nest 8
+              factsDoc
+
+showRuleName :: ProtoRuleName -> String
+showRuleName FreshRule = "rFresh"
+showRuleName (StandRule s) = "r" ++ s
+
+translateRule :: (HighlightDocument d) => [LNFact] -> [LNFact] -> [LNFact] -> M.Map (String, String) String -> (d, M.Map (String, String) String)
+translateRule prems acts concls destrs = 
+    let (docs1, vars1, vars1', destr1) = translatePatterns prems "GET" patternGetsFilter S.empty M.empty destrs
+        (docs2, vars2) = translateNonPatterns prems "GET" nonPatternGetsFilter vars1
+        (docs3, vars3, _, destr3) = translatePatterns prems "IN" patternInsFilter vars2 vars1' destr1
+        (docs4, vars4) = translateNonPatterns prems "IN" nonPatternInsFilter vars3
+        (docs5, vars5) = translateNonPatterns prems "NEW" newsFilter vars4
+        (docs6, vars6) = translateNonPatterns acts "EVENT" (\x -> True) vars5
+        (docs7, vars7) = translateNonPatterns (concls \\ prems) "INSERT" isStorage vars6
+        (docs8, _) = translateNonPatterns concls "OUT" outsFilter vars7
+      in
+    (combineRuleDocs (docs1++docs2++docs3) (docs4++docs5++docs6++docs7++docs8), destr3)
+
+combineRuleDocs :: (HighlightDocument d) => [d] -> [d] -> d
+combineRuleDocs rd1 rd2 = case rd2 of
+                           [] -> vcat rd1 $-$ text "0."
+                           _  -> vcat rd1 $-$ separateRuleDocs rd2
+                          where
+                            separateRuleDocs [r] = r <> text "."
+                            separateRuleDocs (r:rs) = r <> semi $-$ separateRuleDocs rs
+
+
+isStorage :: LNFact -> Bool
+isStorage (Fact tag _ _) = case factTagName tag of
+  "Fr"  -> False
+  "In"  -> False
+  "Out" -> False
+  _     -> True
+
+isPattern :: Term l -> Bool
+isPattern t = case viewTerm t of
+    Lit _ -> False
+    _     -> True
+
+hasPattern :: LNFact -> Bool
+hasPattern (Fact _ _ ts) = 
+  foldl (\acc t -> acc || isPattern t) False ts
+
+patternGetsFilter :: LNFact -> Bool
+patternGetsFilter p = (isStorage p) && hasPattern p
+
+nonPatternGetsFilter :: LNFact -> Bool
+nonPatternGetsFilter p = (isStorage p) && not (hasPattern p)
+
+patternInsFilter :: LNFact -> Bool
+patternInsFilter p@(Fact tag _ _) = (factTagName tag) == "In" && hasPattern p
+
+nonPatternInsFilter :: LNFact -> Bool
+nonPatternInsFilter p@(Fact tag _ _) = (factTagName tag) == "In" && not (hasPattern p)
+
+newsFilter :: LNFact -> Bool
+newsFilter (Fact tag _ _) = (factTagName tag) == "Fr"
+
+outsFilter :: LNFact -> Bool
+outsFilter (Fact tag _ _) = (factTagName tag) == "Out"
+
+translatePatterns :: (HighlightDocument d) => [LNFact] -> String -> (LNFact -> Bool) -> S.Set String -> M.Map String String -> M.Map (String, String) String -> ([d], S.Set String, M.Map String String, M.Map (String, String) String)
+translatePatterns facts factType filterFunction vars helperVars destructors =
```

Comment:
I tried my best to comment it properly in the new commit, hope this makes it more clear now. Some of the comments might be a bit too much, but there happens quite a lot in that function that needs to be explained. Also renamed some variable names which were slightly inaccurate.

I also added a comment to `makeDestructorName` because in that function it also might be hard to understand why I did it things the way I did.